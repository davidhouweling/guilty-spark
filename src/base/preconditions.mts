export const Preconditions = {
  checkExists<T>(value: T, errorMessage?: string): NonNullable<T> {
    if (value == null) {
      throw new Error(errorMessage ?? "Value cannot be null");
    }

    return value;
  },
};
