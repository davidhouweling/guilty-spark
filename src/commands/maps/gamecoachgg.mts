export const GAMECOACH_GG_URLS: Record<string, string> = {
  Aquarius: "https://gamecoach.gg/esports/haloinfinite/aquarius",
  Argyle: "https://gamecoach.gg/esports/haloinfinite/argyle",
  Empyrean: "https://gamecoach.gg/esports/haloinfinite/empyrean",
  Forbidden: "https://gamecoach.gg/esports/haloinfinite/forbidden",
  Fortress: "https://gamecoach.gg/esports/haloinfinite/fortress",
  Fuel: "https://gamecoach.gg/esports/haloinfinite/fuel",
  Inquisitor: "https://gamecoach.gg/esports/haloinfinite/inquisitor",
  Interference: "https://gamecoach.gg/esports/haloinfinite/interference",
  Lattice: "https://gamecoach.gg/esports/haloinfinite/lattice",
  "Live Fire": "https://gamecoach.gg/esports/haloinfinite/livefire",
  Origin: "https://gamecoach.gg/esports/haloinfinite/origin",
  Recharge: "https://gamecoach.gg/esports/haloinfinite/recharge",
  Solitude: "https://gamecoach.gg/esports/haloinfinite/solitude",
  Streets: "https://gamecoach.gg/esports/haloinfinite/streets",
};
