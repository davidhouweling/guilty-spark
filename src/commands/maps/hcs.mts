// HCS map/mode pool

export type MapMode = "Slayer" | "Capture the Flag" | "Strongholds" | "Oddball" | "King of the Hill" | "Neutral Bomb";

export const CURRENT_HCS_MAPS: Record<MapMode, string[]> = {
  Slayer: ["Aquarius", "Live Fire", "Origin", "Recharge", "Solitude", "Streets"],
  "Capture the Flag": ["Aquarius", "Forbidden", "Fortress", "Origin"],
  Strongholds: ["Live Fire", "Lattice", "Recharge"],
  Oddball: ["Live Fire", "Recharge", "Lattice"],
  "King of the Hill": ["Live Fire", "Recharge", "Lattice"],
  "Neutral Bomb": [],
};

export const OBJECTIVE_MODES: Omit<MapMode, "Slayer">[] = [
  "Capture the Flag",
  "Strongholds",
  "Oddball",
  "King of the Hill",
  "Neutral Bomb",
];

export const HCS_SET_FORMAT: Record<number, ("random" | "objective" | "slayer")[]> = {
  1: ["random"],
  3: ["objective", "slayer", "objective"],
  5: ["objective", "slayer", "objective", "objective", "slayer"],
  7: ["objective", "slayer", "objective", "objective", "slayer", "objective", "slayer"],
};

export const HISTORICAL_HCS_MAPS: Record<MapMode, string[]> = {
  Slayer: ["Aquarius", "Live Fire", "Recharge", "Streets", "Solitude", "Empyrean", "Origin"],
  "Capture the Flag": ["Aquarius", "Bazaar", "Empyrean", "Forbidden", "Fortress", "Origin"],
  Strongholds: ["Live Fire", "Recharge", "Streets", "Empyrean", "Solitude", "Lattice"],
  Oddball: ["Live Fire", "Recharge", "Streets", "Lattice"],
  "King of the Hill": ["Live Fire", "Recharge", "Streets", "Solitude", "Lattice"],
  "Neutral Bomb": ["Aquarius", "Fortress"],
};
