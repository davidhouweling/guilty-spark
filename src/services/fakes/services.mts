import { aFakeEnvWith } from "../../base/fakes/env.fake.mjs";
import { aFakeDatabaseServiceWith } from "../database/fakes/database.fake.mjs";
import { aFakeDiscordServiceWith } from "../discord/fakes/discord.fake.mjs";
import { aFakeHaloServiceWith } from "../halo/fakes/halo.fake.mjs";
import type { Services } from "../install.mjs";
import { aFakeXboxServiceWith } from "../xbox/fakes/xbox.fake.mjs";

export function installFakeServicesWith(opts: Partial<Services & { env: Env }> = {}): Services {
  const env = opts.env ?? aFakeEnvWith();
  const databaseService = opts.databaseService ?? aFakeDatabaseServiceWith({ env });
  const discordService = opts.discordService ?? aFakeDiscordServiceWith({ env });
  const xboxService = opts.xboxService ?? aFakeXboxServiceWith({ env });
  const haloService = opts.haloService ?? aFakeHaloServiceWith({ databaseService });

  return {
    databaseService,
    discordService,
    xboxService,
    haloService,
  };
}
