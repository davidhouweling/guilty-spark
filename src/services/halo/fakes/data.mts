import { MapAsset, MatchStats, PlayerMatchHistory } from "halo-infinite-api";
import { readFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

async function readMatchStats(filename: string): Promise<[string, MatchStats]> {
  try {
    const fileContents = await readFile(path.join(__dirname, "data", filename), "utf-8");
    const matchStats = JSON.parse(fileContents) as MatchStats;
    return [matchStats.MatchId, matchStats];
  } catch (error) {
    console.error(`Failed to read match stats from ${filename}: ${error as Error}`);
    throw error;
  }
}

async function readPlayerMatches(): Promise<PlayerMatchHistory[]> {
  try {
    const fileContents = await readFile(path.join(__dirname, "data", "player-matches.json"), "utf-8");
    return JSON.parse(fileContents) as PlayerMatchHistory[];
  } catch (error) {
    console.error(`Failed to read player matches: ${error as Error}`);
    throw error;
  }
}

async function readAssetVersion(): Promise<MapAsset> {
  try {
    const fileContents = await readFile(path.join(__dirname, "data", "asset-version.json"), "utf-8");
    return JSON.parse(fileContents) as MapAsset;
  } catch (error) {
    console.error(`Failed to read asset version: ${error as Error}`);
    throw error;
  }
}

export const matchStats = new Map<string, MatchStats>(
  await Promise.all([
    readMatchStats("ctf.json"),
    readMatchStats("koth.json"),
    readMatchStats("land-grab.json"),
    readMatchStats("oddball.json"),
    readMatchStats("slayer.json"),
    readMatchStats("strongholds.json"),
    readMatchStats("total-control.json"),
  ]),
);

export const playerMatches = await readPlayerMatches();

export const assetVersion = await readAssetVersion();

export const playerXuidsToGametags = new Map([
  ["0100000000000000", "gamertag01"],
  ["0200000000000000", "gamertag02"],
  ["0300000000000000", "gamertag03"],
  ["0400000000000000", "gamertag04"],
  ["0500000000000000", "gamertag05"],
  ["0600000000000000", "gamertag06"],
  ["0700000000000000", "gamertag07"],
  ["0800000000000000", "gamertag08"],
  ["0900000000000000", "gamertag09"],
  ["1000000000000000", "gamertag10"],
  ["1100000000000000", "gamertag11"],
  ["1200000000000000", "gamertag12"],
  ["1300000000000000", "gamertag13"],
  ["1400000000000000", "gamertag14"],
  ["1500000000000000", "gamertag15"],
  ["1600000000000000", "gamertag16"],
]);
