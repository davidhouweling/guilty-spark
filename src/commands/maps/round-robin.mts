import { Preconditions } from "../../base/preconditions.mjs";
import type { MapMode } from "./hcs.mjs";

export interface RoundRobinArgs {
  count: number;
  pool: { mode: MapMode; map: string }[];
  formatSequence: ("slayer" | "objective")[];
}

export type generateRoundRobinMapsFn = (args: RoundRobinArgs) => { mode: MapMode; map: string }[];

export const generateRoundRobinMaps: generateRoundRobinMapsFn = ({ count, pool, formatSequence }) => {
  const available: { mode: MapMode; map: string }[] = [...pool];

  // Shuffle pool for randomness
  for (let i = available.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    const temp: { mode: MapMode; map: string } = Preconditions.checkExists(available[i]);
    available[i] = Preconditions.checkExists(available[j]);
    available[j] = temp;
  }

  const result: { mode: MapMode; map: string }[] = [];
  const used = new Set<string>();
  let poolIndex = 0;

  for (let i = 0; i < count; i++) {
    const type = formatSequence[i % formatSequence.length];
    let candidates: { mode: MapMode; map: string }[] =
      type === "slayer"
        ? available.filter(({ mode }) => mode === "Slayer")
        : available.filter(({ mode }) => mode !== "Slayer");

    candidates = candidates.filter(({ mode, map }) => !used.has(`${String(mode)}:${map}`));

    if (candidates.length === 0) {
      used.clear();
      candidates =
        type === "slayer"
          ? available.filter(({ mode }) => mode === "Slayer")
          : available.filter(({ mode }) => mode !== "Slayer");
    }

    const pick: { mode: MapMode; map: string } = Preconditions.checkExists(candidates[poolIndex % candidates.length]);
    result.push(pick);
    used.add(`${String(pick.mode)}:${pick.map}`);
    poolIndex++;
  }

  return result;
};
