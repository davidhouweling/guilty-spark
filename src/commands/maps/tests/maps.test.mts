import type { MockInstance } from "vitest";
import { describe, it, beforeEach, expect, vi } from "vitest";
import type {
  APIInteractionResponseChannelMessageWithSource,
  APIApplicationCommandInteraction,
  APIMessageComponentButtonInteraction,
  APIInteractionResponseCallbackData,
  APIActionRowComponent,
  APIMessageTopLevelComponent,
  APIButtonComponentWithCustomId,
  APIMessageComponentSelectMenuInteraction,
  APIStringSelectComponent,
  APIEmbed,
  APIMessage,
} from "discord-api-types/v10";
import {
  ApplicationCommandOptionType,
  ApplicationCommandType,
  InteractionType,
  Locale,
  ComponentType,
  MessageFlags,
  ButtonStyle,
  InteractionResponseType,
} from "discord-api-types/v10";
import type { MapMode } from "../hcs.mjs";
import { MapsCommand, PlaylistType, InteractionComponent, FormatType } from "../maps.mjs";
import { installFakeServicesWith } from "../../../services/fakes/services.mjs";
import type { Services } from "../../../services/install.mjs";
import { aFakeEnvWith } from "../../../base/fakes/env.fake.mjs";
import { apiMessage, fakeBaseAPIApplicationCommandInteraction } from "../../../services/discord/fakes/data.mjs";
import { Preconditions } from "../../../base/preconditions.mjs";
import type { generateRoundRobinMapsFn, RoundRobinArgs } from "../round-robin.mjs";

function aFakeMapsInteractionWith(
  options: { name: string; value: unknown; type: number }[] = [],
): APIApplicationCommandInteraction {
  return {
    ...fakeBaseAPIApplicationCommandInteraction,
    type: InteractionType.ApplicationCommand,
    guild: {
      ...fakeBaseAPIApplicationCommandInteraction.guild,
      id: "fake-guild-id",
      locale: Locale.EnglishUS,
      features: [], // Ensure features is always present
    },
    guild_id: "fake-guild-id",
    data: {
      id: "fake-cmd-id",
      name: "maps",
      options,
      resolved: {},
      type: ApplicationCommandType.ChatInput,
    },
  };
}

function getButtonRow(
  components: APIMessageTopLevelComponent[] | undefined,
): APIActionRowComponent<APIButtonComponentWithCustomId> {
  return components?.find(
    (c) =>
      c.type === ComponentType.ActionRow &&
      c.components.every((subComponent) => subComponent.type === ComponentType.Button),
  ) as APIActionRowComponent<APIButtonComponentWithCustomId>;
}

function getButtonById(
  row: APIActionRowComponent<APIButtonComponentWithCustomId> | undefined,
  custom_id: string,
): APIButtonComponentWithCustomId {
  return Preconditions.checkExists(row?.components.find((b) => b.custom_id === custom_id));
}

function getSelectMenu(components: APIMessageTopLevelComponent[] | undefined): APIStringSelectComponent {
  for (const row of components ?? []) {
    if (row.type === ComponentType.ActionRow && "components" in row && Array.isArray(row.components)) {
      for (const c of row.components) {
        if (c.type === ComponentType.StringSelect) {
          return c;
        }
      }
    }
  }

  throw new Error("No select menu found in components");
}

function aFakeMapsMessage({
  playlist = PlaylistType.HcsCurrent,
  format = FormatType.Hcs,
  count = 5,
  selectedPlaylist,
  selectedFormat,
}: {
  playlist?: PlaylistType;
  format?: string;
  count?: 1 | 3 | 5 | 7;
  selectedPlaylist?: PlaylistType | undefined;
  selectedFormat?: string | undefined;
}): {
  embeds: APIEmbed[];
  components: APIMessageTopLevelComponent[];
} {
  const lines = Array(count)
    .fill(0)
    .map((_, i) => String(i + 1));
  return {
    embeds: [
      {
        title: `Maps: ${format}`,
        fields: [
          { name: "#", value: lines.join("\n"), inline: true },
          { name: "Mode", value: Array(count).fill("Slayer").join("\n"), inline: true },
          { name: "Map", value: Array(count).fill("Live Fire").join("\n"), inline: true },
        ],
      } as APIEmbed,
    ],
    components: [
      {
        type: ComponentType.ActionRow,
        components: [
          {
            type: ComponentType.Button,
            custom_id: InteractionComponent.Roll1,
            style: count === 1 ? ButtonStyle.Primary : ButtonStyle.Secondary,
          },
          {
            type: ComponentType.Button,
            custom_id: InteractionComponent.Roll3,
            style: count === 3 ? ButtonStyle.Primary : ButtonStyle.Secondary,
          },
          {
            type: ComponentType.Button,
            custom_id: InteractionComponent.Roll5,
            style: count === 5 ? ButtonStyle.Primary : ButtonStyle.Secondary,
          },
          {
            type: ComponentType.Button,
            custom_id: InteractionComponent.Roll7,
            style: count === 7 ? ButtonStyle.Primary : ButtonStyle.Secondary,
          },
        ],
      } as APIMessageTopLevelComponent,
      {
        type: ComponentType.ActionRow,
        components: [
          {
            type: ComponentType.StringSelect,
            custom_id: InteractionComponent.PlaylistSelect,
            options: [
              {
                label: PlaylistType.HcsCurrent,
                value: PlaylistType.HcsCurrent,
                default: String(selectedPlaylist ?? playlist) === String(PlaylistType.HcsCurrent),
              },
              {
                label: `${PlaylistType.HcsHistorical} (all maps + modes played in any HCS major)`,
                value: PlaylistType.HcsHistorical,
                default: String(selectedPlaylist ?? playlist) === String(PlaylistType.HcsHistorical),
              },
            ],
            placeholder: "Select a playlist",
          },
        ],
      } as APIMessageTopLevelComponent,
      {
        type: ComponentType.ActionRow,
        components: [
          {
            type: ComponentType.StringSelect,
            custom_id: InteractionComponent.FormatSelect,
            options: [
              {
                label: FormatType.Hcs,
                value: FormatType.Hcs,
                default: String(selectedFormat ?? format) === String(FormatType.Hcs),
              },
              {
                label: FormatType.Random,
                value: FormatType.Random,
                default: String(selectedFormat ?? format) === String(FormatType.Random),
              },
              {
                label: FormatType.RandomObjective,
                value: FormatType.RandomObjective,
                default: String(selectedFormat ?? format) === String(FormatType.RandomObjective),
              },
              {
                label: FormatType.RandomSlayer,
                value: FormatType.RandomSlayer,
                default: String(selectedFormat ?? format) === String(FormatType.RandomSlayer),
              },
            ],
          },
        ],
      } as APIMessageTopLevelComponent,
    ],
  };
}

function aFakeApiMessage({
  playlist = PlaylistType.HcsCurrent,
  format = FormatType.Hcs,
  count = 5,
  selectedPlaylist,
  selectedFormat,
}: {
  playlist?: PlaylistType;
  format?: string;
  count?: 1 | 3 | 5 | 7;
  selectedPlaylist?: PlaylistType | undefined;
  selectedFormat?: string | undefined;
}): APIMessage {
  return {
    ...apiMessage,
    embeds: aFakeMapsMessage({ playlist, format, count, selectedPlaylist, selectedFormat }).embeds,
    components: aFakeMapsMessage({ playlist, format, count, selectedPlaylist, selectedFormat }).components,
  };
}

function aFakeButtonInteraction(
  customId: string,
  playlist: PlaylistType = PlaylistType.HcsCurrent,
  format: string = FormatType.Hcs,
  count: 1 | 3 | 5 | 7 = 5,
): APIMessageComponentButtonInteraction {
  return {
    ...fakeBaseAPIApplicationCommandInteraction,
    id: "fake-interaction-id",
    type: InteractionType.MessageComponent,
    data: {
      component_type: ComponentType.Button,
      custom_id: customId,
    },
    message: aFakeApiMessage({ playlist, format, count }),
  };
}

function aFakePlaylistSelectInteraction(
  selectedPlaylist: PlaylistType,
  count: 1 | 3 | 5 | 7 = 5,
  playlist: PlaylistType = PlaylistType.HcsCurrent,
  format: string = FormatType.Hcs,
): APIMessageComponentSelectMenuInteraction {
  return {
    ...fakeBaseAPIApplicationCommandInteraction,
    id: "fake-interaction-id",
    type: InteractionType.MessageComponent,
    data: {
      component_type: ComponentType.StringSelect,
      custom_id: InteractionComponent.PlaylistSelect,
      values: [selectedPlaylist],
    },
    message: aFakeApiMessage({ playlist, format, count, selectedPlaylist }),
  };
}

describe("MapsCommand", () => {
  let command: MapsCommand;
  let services: Services;
  const env = aFakeEnvWith();
  let mockRoundRobin: MockInstance<generateRoundRobinMapsFn>;

  beforeEach(() => {
    services = installFakeServicesWith({ env });
    mockRoundRobin = vi.fn<generateRoundRobinMapsFn>(() => [
      { mode: "Slayer" as MapMode, map: "Live Fire" },
      { mode: "Oddball" as MapMode, map: "Streets" },
      { mode: "Strongholds" as MapMode, map: "Bazaar" },
      { mode: "CTF" as MapMode, map: "Aquarius" },
      { mode: "Slayer" as MapMode, map: "Recharge" },
    ]);
    command = new MapsCommand(services, env, mockRoundRobin as unknown as generateRoundRobinMapsFn);
    vi.spyOn(services.logService, "error").mockImplementation(() => undefined);
    vi.spyOn(services.discordService, "getEmojiFromName").mockReturnValue(":GameCoachGG:");
  });

  describe("/maps basic usage", () => {
    it("renders the mock maps in the embed", () => {
      const interaction = aFakeMapsInteractionWith();
      const { response } = command.execute(interaction);
      const { data } = response as APIInteractionResponseChannelMessageWithSource;
      expect(data).toHaveProperty("embeds");

      const embed = data.embeds?.[0];
      expect(embed).toBeDefined();
      expect(embed?.fields?.length).toBeGreaterThanOrEqual(3);

      const modes = embed?.fields?.[1]?.value.split("\n") ?? [];
      const maps = embed?.fields?.[2]?.value.split("\n") ?? [];
      expect(modes).toEqual(["Slayer", "Oddball", "Strongholds", "CTF", "Slayer"]);
      expect(maps).toEqual([
        "[Live Fire :GameCoachGG:](https://gamecoach.gg/esports/haloinfinite/livefire)",
        "[Streets :GameCoachGG:](https://gamecoach.gg/esports/haloinfinite/streets)",
        "Bazaar",
        "[Aquarius :GameCoachGG:](https://gamecoach.gg/esports/haloinfinite/aquarius)",
        "[Recharge :GameCoachGG:](https://gamecoach.gg/esports/haloinfinite/recharge)",
      ]);
    });
  });

  describe("/maps playlist option", () => {
    it("returns maps from the historical playlist when selected", () => {
      const interaction = aFakeMapsInteractionWith([
        { name: "playlist", value: PlaylistType.HcsHistorical, type: ApplicationCommandOptionType.String },
      ]);
      const { response } = command.execute(interaction);
      const { data } = response as APIInteractionResponseChannelMessageWithSource;
      const embed = data.embeds?.[0];

      expect(embed).toBeDefined();
      expect(embed?.title).toContain(PlaylistType.HcsHistorical);
      expect(embed?.fields?.[0]?.value.split("\n")).toHaveLength(5);
    });

    it("returns maps from the current playlist by default", () => {
      const interaction = aFakeMapsInteractionWith();
      const { response } = command.execute(interaction);
      const { data } = response as APIInteractionResponseChannelMessageWithSource;
      const embed = data.embeds?.[0];

      expect(embed).toBeDefined();
      expect(embed?.title).toContain(PlaylistType.HcsCurrent);
      expect(embed?.fields?.[0]?.value.split("\n")).toHaveLength(5);
    });
  });

  describe("/maps button interaction", () => {
    it("regenerates maps with correct count and playlist when button is pressed", () => {
      const interaction = aFakeButtonInteraction(InteractionComponent.Roll3, PlaylistType.HcsHistorical);
      const { response } = command.execute(interaction);
      const { data } = response as APIInteractionResponseChannelMessageWithSource;
      const embed = data.embeds?.[0];

      expect(embed).toBeDefined();
      expect(embed?.title).toBeDefined();
      expect(embed?.title ?? "").toContain(PlaylistType.HcsHistorical);
      expect(embed?.fields?.[0]?.value.split("\n")).toHaveLength(5);

      const buttonRowHist = getButtonRow(data.components);
      expect(getButtonById(buttonRowHist, InteractionComponent.Roll3).style).toBe(ButtonStyle.Primary);
    });

    it("throws for unknown button id", () => {
      const interaction = aFakeButtonInteraction("btn_maps_roll_unknown");
      const { response } = command.execute(interaction);
      const { data } = response as APIInteractionResponseChannelMessageWithSource;

      expect(data.content).toMatch(/Unreachable code with specified value/);
      expect(data.flags).toBe(MessageFlags.Ephemeral);
    });

    it("throws for unknown playlist in embed", () => {
      const interaction = aFakeButtonInteraction(InteractionComponent.Roll1, "NotAPlaylist" as PlaylistType);
      const { response } = command.execute(interaction);
      const { data } = response as APIInteractionResponseChannelMessageWithSource;

      expect(typeof data.content).toBe("string");
      expect(data.content).toMatch(/Playlist not found/i);
      expect(data.flags).toBe(MessageFlags.Ephemeral);
    });

    it("returns DeferredMessageUpdate and jobToComplete for Initiate button", async () => {
      const interaction = aFakeButtonInteraction(InteractionComponent.Initiate);
      const { response, jobToComplete } = command.execute(interaction);

      expect(response).toEqual({ type: InteractionResponseType.DeferredMessageUpdate });
      expect(jobToComplete).toBeDefined();

      const createMessageSpy = vi.spyOn(services.discordService, "createMessage").mockResolvedValue(apiMessage);
      await jobToComplete?.();
      const [channelId, data] = createMessageSpy.mock.calls[0] as [string, APIInteractionResponseCallbackData];

      expect(channelId).toBe(interaction.channel.id);
      expect(data.embeds?.[0]?.title).toContain("Maps: HCS - current");

      const actionRow = getButtonRow(data.components);
      expect(actionRow.components).toHaveLength(4);
    });
  });

  describe("/maps repost button interaction", () => {
    it("returns DeferredMessageUpdate and reposts the maps, then deletes the original message", async () => {
      const interaction = aFakeButtonInteraction(InteractionComponent.Repost);
      const { response, jobToComplete } = command.execute(interaction);
      expect(response).toEqual({ type: InteractionResponseType.DeferredMessageUpdate });
      expect(jobToComplete).toBeDefined();

      const createMessageSpy = vi.spyOn(services.discordService, "createMessage").mockResolvedValue(apiMessage);
      const deleteMessageSpy = vi.spyOn(services.discordService, "deleteMessage").mockResolvedValue();
      await jobToComplete?.();

      expect(createMessageSpy).toHaveBeenCalledTimes(1);
      expect(deleteMessageSpy).toHaveBeenCalledTimes(1);

      const [channelId, data] = createMessageSpy.mock.calls[0] as [string, APIInteractionResponseCallbackData];
      expect(channelId).toBe(interaction.channel.id);
      expect(data.embeds?.[0]?.title).toContain("Maps: HCS - current");

      const [delChannelId, delMessageId, delReason] = deleteMessageSpy.mock.calls[0] as [string, string, string];
      expect(delChannelId).toBe(interaction.channel.id);
      expect(delMessageId).toBe(interaction.message.id);
      expect(delReason).toBe("Reposting maps");
    });
  });

  describe("/maps playlist select interaction", () => {
    it("updates the embed and buttons when playlist is switched to historical", () => {
      const interaction = aFakePlaylistSelectInteraction(PlaylistType.HcsHistorical, 3, PlaylistType.HcsHistorical);
      const { response } = command.execute(interaction);
      const { data } = response as APIInteractionResponseChannelMessageWithSource;
      const embed = data.embeds?.[0];

      expect(embed).toBeDefined();
      expect(typeof embed?.title).toBe("string");
      expect(embed?.title ?? "").toContain(PlaylistType.HcsHistorical);

      const buttonRow = getButtonRow(data.components);
      expect(getButtonById(buttonRow, InteractionComponent.Roll3).style).toBe(ButtonStyle.Primary);

      const select = getSelectMenu(data.components);
      expect(select.custom_id).toBe(InteractionComponent.PlaylistSelect);
      expect(select.options.find((o) => o.value === String(PlaylistType.HcsHistorical))?.default).toBe(true);
      expect(select.options.find((o) => o.value === String(PlaylistType.HcsCurrent))?.default).toBe(false);
    });

    it("updates the embed and buttons when playlist is switched to current", () => {
      const interaction = aFakePlaylistSelectInteraction(PlaylistType.HcsCurrent, 7, PlaylistType.HcsCurrent);
      const { response } = command.execute(interaction);
      const { data } = response as APIInteractionResponseChannelMessageWithSource;
      const embed = data.embeds?.[0];

      expect(embed).toBeDefined();
      expect(embed?.title).toBeDefined();
      expect(embed?.title).toContain(PlaylistType.HcsCurrent);

      const buttonRow = getButtonRow(data.components);
      expect(getButtonById(buttonRow, InteractionComponent.Roll7).style).toBe(ButtonStyle.Primary);

      const select = getSelectMenu(data.components);
      expect(select.custom_id).toBe(InteractionComponent.PlaylistSelect);
      expect(select.options.find((o) => o.value === String(PlaylistType.HcsCurrent))?.default).toBe(true);
      expect(select.options.find((o) => o.value === String(PlaylistType.HcsHistorical))?.default).toBe(false);
    });
  });

  it("calls the injected roundRobinFn with correct arguments", () => {
    command.generateMaps({
      count: 5,
      playlist: PlaylistType.HcsCurrent,
      format: FormatType.Hcs,
    });
    expect(mockRoundRobin.mock.calls.length).toBe(1);

    const [call] = mockRoundRobin.mock.calls[0] as [RoundRobinArgs];
    expect(call.count).toBe(5);
    expect(Array.isArray(call.pool)).toBe(true);
    expect(Array.isArray(call.formatSequence)).toBe(true);
  });
});
