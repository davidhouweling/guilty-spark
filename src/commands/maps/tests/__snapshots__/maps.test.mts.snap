// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MapsCommand > execute(): application command > jobToComplete > calls discordService.updateDeferredReply with maps embed 1`] = `
[
  "fake-token",
  {
    "components": [
      {
        "components": [
          {
            "custom_id": "btn_maps_roll_1",
            "label": "Regen maps (count: 1)",
            "style": 2,
            "type": 2,
          },
          {
            "custom_id": "btn_maps_roll_3",
            "label": "Regen maps (count: 3)",
            "style": 2,
            "type": 2,
          },
          {
            "custom_id": "btn_maps_roll_5",
            "label": "Regen maps (count: 5)",
            "style": 1,
            "type": 2,
          },
          {
            "custom_id": "btn_maps_roll_7",
            "label": "Regen maps (count: 7)",
            "style": 2,
            "type": 2,
          },
        ],
        "type": 1,
      },
      {
        "components": [
          {
            "custom_id": "select_maps_playlist",
            "options": [
              {
                "default": true,
                "description": "The current maps and modes of HCS (as of 26 August 2025)",
                "label": "HCS - Current",
                "value": "C",
              },
              {
                "default": false,
                "description": "All maps and modes that have been played at any HCS major event",
                "label": "HCS - Historical",
                "value": "H",
              },
              {
                "default": false,
                "description": "Ranked Arena playlist",
                "label": "Ranked Arena",
                "value": "R",
              },
              {
                "default": false,
                "description": "Ranked Slayer playlist",
                "label": "Ranked Slayer",
                "value": "S",
              },
              {
                "default": false,
                "description": "Ranked Snipers playlist",
                "label": "Ranked Snipers",
                "value": "N",
              },
              {
                "default": false,
                "description": "Ranked Tactical playlist",
                "label": "Ranked Tactical",
                "value": "T",
              },
              {
                "default": false,
                "description": "Ranked Doubles playlist",
                "label": "Ranked Doubles",
                "value": "D",
              },
              {
                "default": false,
                "description": "Ranked FFA playlist",
                "label": "Ranked FFA",
                "value": "F",
              },
              {
                "default": false,
                "description": "Ranked Squad Battle playlist",
                "label": "Ranked Squad Battle",
                "value": "Q",
              },
            ],
            "placeholder": "Select a playlist",
            "type": 3,
          },
        ],
        "type": 1,
      },
      {
        "components": [
          {
            "custom_id": "select_maps_format",
            "options": [
              {
                "default": true,
                "description": "Obj, slayer, obj, obj, slayer, ...",
                "label": "HCS",
                "value": "H",
              },
              {
                "default": false,
                "description": "Randomly pick objective or slayer for each map",
                "label": "Random",
                "value": "R",
              },
              {
                "default": false,
                "description": "Only pick objective modes",
                "label": "Objective only",
                "value": "O",
              },
              {
                "default": false,
                "description": "Only pick slayer modes",
                "label": "Slayer only",
                "value": "S",
              },
            ],
            "type": 3,
          },
        ],
        "type": 1,
      },
      {
        "components": [
          {
            "custom_id": "btn_maps_repost",
            "emoji": {
              "name": "‚è¨",
            },
            "label": "Move to bottom of chat",
            "style": 2,
            "type": 2,
          },
        ],
        "type": 1,
      },
    ],
    "embeds": [
      {
        "color": 5793266,
        "fields": [
          {
            "inline": true,
            "name": "#",
            "value": "1
2
3
4
5",
          },
          {
            "inline": true,
            "name": "Mode",
            "value": "Slayer
Oddball
Strongholds
CTF
Slayer",
          },
          {
            "inline": true,
            "name": "Map",
            "value": "[Live Fire :GameCoachGG:](https://gamecoach.gg/esports/haloinfinite/livefire)
[Streets :GameCoachGG:](https://gamecoach.gg/esports/haloinfinite/streets)
Bazaar
[Aquarius :GameCoachGG:](https://gamecoach.gg/esports/haloinfinite/aquarius)
[Recharge :GameCoachGG:](https://gamecoach.gg/esports/haloinfinite/recharge)",
          },
          {
            "name": "",
            "value": "-# Generated by <@discord_user_01>",
          },
        ],
        "title": "Maps: HCS - Current",
      },
    ],
  },
]
`;
